packages/tiny-agents/src/lib/webServer.ts

```diff
@@ -0,0 +1,129 @@
+import type { IncomingMessage } from "node:http";
+import { createServer, ServerResponse } from "node:http";
+import type { AddressInfo } from "node:net";
+import { z } from "zod";
+import type { Agent } from "../index";
+import { ANSI } from "./utils";
+import { stdout } from "node:process";
+import type { ChatCompletionStreamOutput } from "@huggingface/tasks";
+
+const REQUEST_ID_HEADER = "X-Request-Id";
+// Generate a simple UUID v4 without dependencies
+const generateUUID = () => {
+	return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
+		const r = (Math.random() * 16) | 0;
+		const v = c === "x" ? r : (r & 0x3) | 0x8;
+		return v.toString(16);
+	});
+};
+const ChatCompletionInputSchema = z.object({
+	messages: z.array(
+		z.object({
+			role: z.enum(["user", "assistant"]),
+			content: z.string(),
+		})
+	),
+	/// Only allow stream: true
+	stream: z.literal(true),
+});
+function getJsonBody(req: IncomingMessage) {
+	return new Promise((resolve, reject) => {
+		let data = "";
+		req.on("data", (chunk) => (data += chunk));
+		req.on("end", () => {
+			try {
+				resolve(JSON.parse(data));
+			} catch (e) {
+				reject(e);
+			}
+		});
+		req.on("error", reject);
+	});
+}
+class ServerResp extends ServerResponse {
+	error(statusCode: number, reason: string) {
+		this.writeHead(statusCode).end(JSON.stringify({ error: reason }));
+	}
+}
+
+export function startServer(agent: Agent): void {
+	const server = createServer({ ServerResponse: ServerResp }, async (req, res) => {
+		res.setHeader(REQUEST_ID_HEADER, generateUUID());
+		res.setHeader("Content-Type", "application/json");
+		if (req.method === "POST" && req.url === "/v1/chat/completions") {
+			let body: unknown;
+			let requestBody: z.infer<typeof ChatCompletionInputSchema>;
+			try {
+				body = await getJsonBody(req);
+			} catch {
+				return res.error(404, "Invalid JSON");
+			}
+			try {
+				requestBody = ChatCompletionInputSchema.parse(body);
+			} catch (err) {
+				if (err instanceof z.ZodError) {
+					return res.error(404, "Invalid ChatCompletionInput body \n" + JSON.stringify(err));
+				}
```

Comment:
Use a recent version of zod, you can `import { z } from "zod/v4"`, and you can use `z.prettifyError(err)`

Reply:
i tried but it requires ESM or something similar "node16" or something.. (but feel free to give it a try)

Reply:
did you want to give it a try @coyotte508? would be cool if we were able to use Zod/v4